WikiClientLibrary.Cargo/CargoWikiSiteExtensions.cs

```diff
@@ -0,0 +1,86 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using Newtonsoft.Json.Linq;
+using WikiClientLibrary.Client;
+using WikiClientLibrary.Infrastructures.Logging;
+using WikiClientLibrary.Sites;
+
+namespace WikiClientLibrary.Cargo
+{
+
+    /// <summary>
+    /// Extension methods for MediaWiki sites with <a href="https://www.mediawiki.org/wiki/Special:MyLanguage/Extension:Cargo">Cargo extension</a> installed.
+    /// </summary>
+    public static class CargoWikiSiteExtensions
+    {
+
+        /// <inheritdoc cref="ExecuteCargoQueryAsync(WikiSite,CargoQueryParameters,CancellationToken)"/>
+        public static Task<IList<JObject>> ExecuteCargoQueryAsync(this WikiSite site, CargoQueryParameters queryParameters)
+        {
+            return ExecuteCargoQueryAsync(site, queryParameters, default);
+        }
+
+        /// <summary>
+        /// Executes a Cargo query and retrieves the response rows.
+        /// </summary>
+        /// <param name="site">a MediaWiki site with Cargo extension installed.</param>
+        /// <param name="queryParameters">query parameters.</param>
+        /// <param name="cancellationToken">a token used to cancel the operation.</param>
+        /// <exception cref="ArgumentNullException"><paramref name="site"/> or <paramref name="queryParameters"/> is <c>null</c>.</exception>
+        /// <exception cref="ArgumentOutOfRangeException">
+        /// <paramref name="queryParameters"/>.<see cref="CargoQueryParameters.Limit" /> is 0 or negative.
+        /// - or -
+        /// <paramref name="queryParameters"/>.<see cref="CargoQueryParameters.Offset" /> is negative.
+        /// </exception>
+        /// <exception cref="MediaWikiRemoteException">
+        /// When query execution failed, Cargo usually fails with <c>internal_api_error_MWException</c> error.
+        /// You should observe <c>"MWException"</c> as the value of <see cref="MediaWikiRemoteException.ErrorClass"/>.
+        /// </exception>
+        /// <returns>a list of rows, each item is a JSON object with field name as key.</returns>
+        public static async Task<IList<JObject>> ExecuteCargoQueryAsync(this WikiSite site, CargoQueryParameters queryParameters, CancellationToken cancellationToken)
+        {
+            if (site == null) throw new ArgumentNullException(nameof(site));
+            if (queryParameters == null) throw new ArgumentNullException(nameof(queryParameters));
+            if (queryParameters.Limit <= 0)
+                throw new ArgumentOutOfRangeException(nameof(queryParameters) + "." + nameof(queryParameters.Limit));
+            if (queryParameters.Offset < 0)
+                throw new ArgumentOutOfRangeException(nameof(queryParameters) + "." + nameof(queryParameters.Offset));
+            if (queryParameters.Tables == null)
```

Comment:
Actually there's one other check you can do, if you want - because of how the response is encoded, none of the field name *aliases* can start with an underscore. So for example if you're querying `_pageName` you would have to alias this as `_pageName=Page`. Additionally, `Champion=_Champion` would be illegal. And Cargo will not provide any kind of helpful api error response if the user fails to alias this way. So you could check for that for every field.

Reply:
Thanks for your reminder! Generally I'd like to make the pre-condition as permissive as enough, and let MW API Server be the single source of truth. Thus for now I will only do the nullability check.

But specifically for Cargo, we may do some postmortem automatically and help user "guess" what could be the cause if we receive DBError / InternalError. This can be covered later.